server/src/core/distribution/run-distribution.ts

```diff
@@ -46,11 +47,19 @@ interface CreateDistributionPlanResult {
   beneficiariesSummaries:DistributionPlanBeneficiariesSummaries;
 }
 
-export async function runDonationDistribution(db: Db, args: DonationDistributionArgs): Promise<DonationDistributionEvent[]> {
+export async function runDonationDistribution(db: Db, args: DonationDistributionArgs, filter: BeneficiaryFilter): Promise<DonationDistributionEvent[]> {
   const { periodLength, periodLimit, users } = args;
 
-  const beneficiaries = await findEligibleBeneficiaries(db, periodLimit, periodLength);
-  const donorIds = beneficiaries.reduce<string[]>((acc, b) => [...acc, ...b.donors], []);
+  const beneficiaries = await findEligibleBeneficiaries(db, periodLimit, periodLength, filter);
+  let donorIds;
+  if (beneficiaries.length && beneficiaries[0].donors) {
+    donorIds = beneficiaries.reduce<string[]>((acc, b) => [...acc, ...b.donors], []);
+  }
+  else {
+    const donors: User[] = await db.collection(USERS_COLL).find({ roles: { $in: ['donor'] } }).toArray();
```

Comment:
I think you should make a projection here to return just the donorids so that we reduce the load on the database since this will potentially return a lot of data.

Reply:
Actually, since `donorIds` are only used to pass the input to `computeDonorsBalances`, which in turn fetches donors from the db, then we can avoid making this call to the db on line 59. We can make it so that the second argument to `computeDonorsBalances` is optional. When it's provided, then it will work the way it currently works, and when it's not provided then the query will exclude the `_id: { $in: donors }` part.

So the `runDonationDistribution` function would be:

```ts
function runDonationDistribution(db: Db, args: DonationDistributionArgs, onlyVettedBeneficiaries: boolean = false): Promise<DonationDistributionEvent[]> {
  const { periodLength, periodLimit, users } = args;

  const beneficiariesFilter = onlyVettedBeneficiaries ? { isVetted: true, beneficiaryStatus: 'verified' } : {};
  const beneficiaries = await findEligibleBeneficiaries(db, periodLimit, periodLength, beneficiariesFilter);

  let donors: DonorBalance[];
  if (onlyVettedBeneficiaries) {
    // vetted beneficiaries receive donations from any donor
    donors = await computeDonorsBalances(db);
  }
  else {
    const donorIds = beneficiaries.reduce<string[]>((acc, b) => [...acc, ...(b.donors || [])], []);
    donors = await computeDonorsBalances(db, Array.from(new Set(donorIds)));
  }

  const plan = createDistributionPlan(beneficiaries, donors);
  const result = await executeDistributionPlan(users, plan.transfers);
  return result;
}
```